File: python/TransformationTasks.py

import base64
import sys
import crypto
sys.modules['Crypto'] = crypto
from crypto.Cipher import DES, DES3, AES
import Pipeline as P

from abc import ABC, abstractmethod

class TransformationTask(ABC):
    keys_counter = 0

    # Static methods

    def set_input_file(input_path):
        file = open(input_path, 'rb')
        TransformationTask.message_bytes = file.read()
        file.close()

    def set_output_file(output_path):
        TransformationTask.output_file = open(output_path, 'w')
        TransformationTask.output_file.write('key,output\n')
    
    def close_output_file():
        TransformationTask.output_file.close()

    # Instance methods

    def run(self):
        while TransformationTask.keys_counter < P.Pipeline.total_keys:
            if not P.Pipeline.key_queue.empty():
                key = P.Pipeline.key_queue.get()
                output_bytes = self._transform_bytes(key)
                key_string = f'{base64.b64encode(key)}'[2:-1]
                output_string = f'{base64.b64encode(output_bytes)}'[2:-1]
                TransformationTask.output_file.write(f'{key_string},{output_string}\n')
                TransformationTask.keys_counter += 1 

    @abstractmethod
    def _transform_bytes(self, key):
        pass


# Concrete Transformation Tasks

# DES

class DESEncryption(TransformationTask):
    def _transform_bytes(self, key):
        cipher = DES.new(key, DES.MODE_OFB,iv=(0).to_bytes(8, 'big'))
        encrypted_msg = cipher.encrypt(TransformationTask.message_bytes)
        return encrypted_msg

class DESDecryption(TransformationTask):
    def _transform_bytes(self, key):
        cipher = DES.new(key, DES.MODE_OFB, iv=(0).to_bytes(8, 'big'))
        decrypted_msg = cipher.decrypt(TransformationTask.message_bytes)
        return decrypted_msg

# DES3

class DES3Encryption(TransformationTask):
    def _transform_bytes(self, key):
        cipher = DES3.new(key, DES3.MODE_CFB, iv=(0).to_bytes(8, 'big'))
        encrypted_msg = cipher.encrypt(TransformationTask.message_bytes)
        return encrypted_msg

class DES3Decryption(TransformationTask):
    def _transform_bytes(self, key):
        cipher = DES3.new(key, DES3.MODE_CFB, iv=(0).to_bytes(8, 'big'))
        decrypted_msg = cipher.decrypt(TransformationTask.message_bytes)
        return decrypted_msg

# AES

class AESEncryption(TransformationTask):
    def _transform_bytes(self, key):
        cipher = AES.new(key, AES.MODE_CFB, iv=(0).to_bytes(16, 'big'))
        encrypted_msg = cipher.encrypt(TransformationTask.message_bytes)
        return encrypted_msg

class AESDecryption(TransformationTask):
    def _transform_bytes(self, key):
        cipher = AES.new(key, AES.MODE_CFB, iv=(0).to_bytes(16, 'big'))
        decrypted_msg = cipher.decrypt(TransformationTask.message_bytes)
        return decrypted_msg

File: java/TransformationTasks.java
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.Base64.Encoder;
import javax.crypto.Cipher;
import javax.crypto.SecretKey;


abstract class TransformationTask implements Runnable{

    private static Integer keyCounter = 0;
    private static FileWriter outputFile;
    protected static byte[] messageBytes;
    protected Cipher cipher;

    public static void setInputFile(String inputPath) throws IOException{
        TransformationTask.messageBytes = Files.readAllBytes(Paths.get(inputPath));
    }

    public static void setOutputFile(String outputPath) throws IOException{
        TransformationTask.outputFile = new FileWriter(outputPath);
        TransformationTask.outputFile.write("key,output\n");
    }

    public static void close_output_file() throws IOException{
        TransformationTask.outputFile.close();
    }

    public static void resetKeyCounter(){
        TransformationTask.keyCounter = 0;
    }

    @Override
    public void run() {
        while (TransformationTask.keyCounter < Pipeline.totalKeys){
            SecretKey key = TransformationTask.get_next_key();
            if(key == null){
                continue;
            }
            byte[] outputBytes = null;
            Encoder encoder = Base64.getEncoder();
            String keyString = encoder.encodeToString(key.getEncoded());
            String outputString;
            try {
                outputBytes = this.transformBytes(key);
                outputString = encoder.encodeToString(outputBytes);
            } catch (Exception e) {
                outputString = "Could not encrypt/decrypt. Wrong Key/Input Format";
            }
            try {
                TransformationTask.outputFile.write(String.format("%s,%s\n", keyString, outputString));
            } catch (IOException e) {}
            TransformationTask.increaseKeyCounter();
        }
    }

    synchronized
    private static SecretKey get_next_key(){
        SecretKey key = null;
        try{
            key = Pipeline.keyQueue.remove();
        }catch(Exception e){}
        return key;
    }

    synchronized
    private static void increaseKeyCounter(){
        TransformationTask.keyCounter  ++;
    }

    abstract protected byte[] transformBytes(SecretKey key) throws Exception;
}


// Concrete Transformation Tasks

// DES

class DESEncryption extends TransformationTask{
    public DESEncryption() throws Exception{
        this.cipher = Cipher.getInstance("DES");
    }
    @Override
    protected byte[] transformBytes(SecretKey key) throws Exception{
        this.cipher.init(Cipher.ENCRYPT_MODE, key);
        byte[] encryptedBytes = this.cipher.doFinal(TransformationTask.messageBytes);
        return encryptedBytes;
    }
}

class DESDecryption extends TransformationTask{
    public DESDecryption() throws Exception{
        this.cipher = Cipher.getInstance("DES");
    }
    @Override
    protected byte[] transformBytes(SecretKey key) throws Exception{
        this.cipher.init(Cipher.DECRYPT_MODE, key);
        byte[] decryptedBytes;
        try{
            decryptedBytes = this.cipher.doFinal(TransformationTask.messageBytes);
        }catch(Exception e){
            decryptedBytes = null;
        }
        return decryptedBytes;
    }
}

// DES3

class DES3Encryption extends TransformationTask{
    public DES3Encryption() throws Exception{
        this.cipher = Cipher.getInstance("DESede");
    }
    @Override
    protected byte[] transformBytes(SecretKey key) throws Exception{
        this.cipher.init(Cipher.ENCRYPT_MODE, key);
        byte[] encryptedBytes = this.cipher.doFinal(TransformationTask.messageBytes);
        return encryptedBytes;
    }
}

class DES3Decryption extends TransformationTask{
    public DES3Decryption() throws Exception{
        this.cipher = Cipher.getInstance("DESede");
    }
    @Override
    protected byte[] transformBytes(SecretKey key) throws Exception{
        this.cipher.init(Cipher.DECRYPT_MODE, key);
        byte[] decryptedBytes;
        try{
            decryptedBytes = this.cipher.doFinal(TransformationTask.messageBytes);
        }catch(Exception e){
            decryptedBytes = null;
        }
        return decryptedBytes;
    }
}

// AES

class AESEncryption extends TransformationTask{
    public AESEncryption() throws Exception{
        this.cipher = Cipher.getInstance("AES");
    }
    @Override
    protected byte[] transformBytes(SecretKey key) throws Exception{
        this.cipher.init(Cipher.ENCRYPT_MODE, key);
        byte[] encryptedBytes = this.cipher.doFinal(TransformationTask.messageBytes);
        return encryptedBytes;
    }
}

class AESDecryption extends TransformationTask{
    public AESDecryption() throws Exception{
        this.cipher = Cipher.getInstance("AES");
    }
    @Override
    protected byte[] transformBytes(SecretKey key) throws Exception{
        this.cipher.init(Cipher.DECRYPT_MODE, key);
        byte[] decryptedBytes;
        try{
            decryptedBytes = this.cipher.doFinal(TransformationTask.messageBytes);
        }catch(Exception e){
            decryptedBytes = null;
        }
        return decryptedBytes;
    }
}